File: src/main/services/calendar/calendar-manager.ts
import { NormalizedEvent } from '../../../shared/types/calendar';
import { appStore } from '../../store/app-store';
import { getICSEvents } from './ics-calendar';

export class CalendarManager {
  private cachedEvents: NormalizedEvent[] = [];
  private lastFetchTime: Date | null = null;

  async fetchAllEvents(hoursBehind = 168, hoursAhead = 336): Promise<NormalizedEvent[]> {
    const allEvents: NormalizedEvent[] = [];
    const errors: string[] = [];

    // Fetch ICS accounts in parallel
    const icsAccounts = appStore.getICSAccounts();
    const icsPromises = icsAccounts.map(async account => {
      try {
        const events = await getICSEvents(account.url, account.id, hoursAhead, hoursBehind);
        allEvents.push(...events);
      } catch (e) {
        errors.push(`ICS (${account.displayName}): ${(e as Error).message}`);
        console.error(`Failed to fetch ICS events for ${account.displayName}:`, e);
      }
    });

    await Promise.all(icsPromises);

    // Sort by start time
    allEvents.sort((a, b) => new Date(a.start).getTime() - new Date(b.start).getTime());

    this.cachedEvents = allEvents;
    this.lastFetchTime = new Date();

    if (errors.length > 0) {
      console.warn('Some calendar fetches failed:', errors);
    }

    return allEvents;
  }

  getCachedEvents(): NormalizedEvent[] {
    return this.cachedEvents;
  }

  getLastFetchTime(): Date | null {
    return this.lastFetchTime;
  }
}

let instance: CalendarManager | null = null;

export function getCalendarManager(): CalendarManager {
  if (!instance) instance = new CalendarManager();
  return instance;
}

File: lib/services/calendar/calendar_manager.dart
import 'dart:math' show min;
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../../models/calendar_event.dart';
import '../../core/time_utils.dart';
import '../storage/app_database.dart';
import 'ics_calendar_service.dart';
import 'outlook_calendar_service.dart';

class CalendarManager {
  static CalendarManager? _instance;
  final IcsCalendarService _icsService = IcsCalendarService();
  final OutlookCalendarService _outlookService = OutlookCalendarService();
  List<NormalizedEvent> _cachedEvents = [];

  CalendarManager._();

  static CalendarManager getInstance() {
    _instance ??= CalendarManager._();
    return _instance!;
  }

  List<NormalizedEvent> get cachedEvents => _cachedEvents;

  Future<List<NormalizedEvent>> fetchAllEvents() async {
    final db = await AppDatabase.getInstance();
    final accounts = db.getAccounts();
    final allEvents = <NormalizedEvent>[];

    for (final account in accounts) {
      if (account.provider != 'ics' || account.icsUrl == null) continue;
      try {
        final events = await _icsService.fetchEvents(account.id, account.icsUrl!);
        allEvents.addAll(events);
      } catch (e) {
        debugPrint('[CalendarManager] Failed to fetch for account ${account.id}: $e');
        // If this is an Outlook-hosted feed that returned an auth-style error
        // (common when a corporate Defender network extension routes traffic
        // through Azure and Exchange Online rejects the anonymous request),
        // fall back to reading the same calendar from the local Outlook app.
        if (_isOutlookUrl(account.icsUrl!) && _isAuthError(e)) {
          debugPrint('[CalendarManager] Auth error on Outlook ICS — trying local Outlook fallback');
          final fallback = await _outlookService.fetchEvents(account.id);
          if (fallback.isNotEmpty) {
            debugPrint('[CalendarManager] Outlook fallback: ${fallback.length} events added');
            allEvents.addAll(fallback);
          }
        }
      }
    }

    allEvents.sort((a, b) => parseToLocal(a.start).compareTo(parseToLocal(b.start)));

    // Deduplicate: same title + same start minute = same meeting across feeds.
    final seen = <String>{};
    final deduped = <NormalizedEvent>[];
    for (final event in allEvents) {
      final key =
          '${event.title.trim().toLowerCase()}|${event.start.substring(0, min(16, event.start.length))}';
      if (seen.add(key)) deduped.add(event);
    }

    _cachedEvents = deduped;
    return deduped;
  }

  // ---------------------------------------------------------------------------
  // Helpers for Outlook fallback detection
  // ---------------------------------------------------------------------------

  /// Returns true if [url] is an Outlook / Exchange Online ICS feed.
  static bool _isOutlookUrl(String url) =>
      url.contains('outlook.office365.com') || url.contains('outlook.office.com');

  /// Returns true if [error] looks like an HTTP auth/rejection error (400, 401,
  /// 403, 407) — the typical codes returned when Exchange Online blocks an
  /// anonymous ICS request from a corporate-managed device.
  static bool _isAuthError(Object error) {
    if (error is DioException) {
      final code = error.response?.statusCode;
      return code == 400 || code == 401 || code == 403 || code == 407;
    }
    return false;
  }
}
